### PR TITLE
safekeeper: don't un-evict timelines during snapshot API handler

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

When we use pull_timeline API on an evicted timeline, it gets downloaded to serve the snapshot API request.  That means that to evacuate all the timelines from a node, the node needs enough disk space to download partial segments from all timelines, which may not be physically the case.

TODO:
- [x] Test of pull_timeline while evicted
- [x] Does the destination of the move go correctly into an evicted state after pulling?
- [x] Is it safe to read the checkpoint of an offloaded timeline without holding some guard that prevents it activating & writing the checkpoint we are reading: is the read atomic by virtue of the mutex we hold?
- [x] In case it is unevicted concurrently, remote S3 object might be mutated: after copying remote object, we should re-check we're still offloaded, to guarantee that control file is consistent with what we copied remotely.
- [x] Reduce duplication between offloaded and regular code paths

Closes: #8833 

## Summary of changes

- Add a "try" variant of acquiring a residence guard, that returns None if the timeline is offloaded
- During snapshot API handler, take a different code path if the timeline isn't resident, where we just read the checkpoint and don't try to read any segments.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: safekeeper/src/pull_timeline.rs
Comment (jcsp) [inline safekeeper/src/pull_timeline.rs:208]: @arssher is the lock acquired in `write_shared_state` sufficient to make it safe to go read the control file?  At this point in the code I have checked that the timeline is offloaded (so it shouldn't be mutating things), but there isn't anything guarding it against getting un-offloaded while this code runs.

Comment (arssher) [reply safekeeper/src/pull_timeline.rs:208]: See `switch_to_present`; WAL file might be concurrently downloaded, but cfile update is guarded.

File: safekeeper/src/pull_timeline.rs
Comment (erikgrinaker) [inline safekeeper/src/pull_timeline.rs:237]: Unlikely, but what would happen if the timeline was un-offloaded, modified, and then offloaded again in the meanwhile? Can we detect this somehow and at least fail safe, e.g. by comparing the control file modification time?

**Review by erikgrinaker — APPROVED** [review]
LGTM modulo the potential race, but I'm not overly familiar with this area yet.

File: safekeeper/src/pull_timeline.rs
Comment (jcsp) [reply safekeeper/src/pull_timeline.rs:237]: It's not clear to me if a timeline can go through a whole offloaded->restore->offload cycle while this function is still  holding WriteGuardSharedState (or if it can, then can the control file change?)

Will wait for Arseny's input

File: test_runner/regress/test_wal_acceptor.py
Comment (arssher) [inline test_runner/regress/test_wal_acceptor.py:2060]: Nit: file removal happens outside timeline lock, so there is a small probability of this assertion violation. May ignore it or turn into waiting.

File: test_runner/regress/test_wal_acceptor.py
Comment (arssher) [inline test_runner/regress/test_wal_acceptor.py:2073]: This is also racy because metric is incremented in the manager which start is not synced with pull_timeline finish.

**Review by arssher — APPROVED** [review]
There are indeed issues with racing: doing `try_wal_residence_guard` while holding timeline lock is bad because timeline manager might be locking the timeline for e.g. uneviction. However, since s3 copy is done also under timeline lock (which is not ideal, but probably ok), it is easy to fix: just lock the state, check that timeline is in offloaded state and do all work while holding the lock. Timeline can't be unevicted while lock is held.

Ideally we could 1) check that timeline is evicted, remember cfile contents including partial seg name 2) do s3 copy, ensure it is good 3) stream the patched cfile contents in the response.

**Review by arssher — COMMENTED** [review]
(apparently I pressed approval button in the review, but too early for that, so resubmitting without approval)

**Review by arssher — CHANGES_REQUESTED** [review]
(apparently I pressed approval button in the review, but too early for that, so resubmitting without approval)

File: safekeeper/src/pull_timeline.rs
Comment (erikgrinaker) [reply safekeeper/src/pull_timeline.rs:237]: Right, I think it would have to happen before we acquire `WriteGuardSharedState` (e.g. if we end up blocking on it for a long time, or under very high scheduling latency). But you're right -- the control file is accessed under lock, and if we're still offloaded under that lock then I think we should be fine (unless I'm missing some more subtle interaction).

File: safekeeper/src/pull_timeline.rs
Comment (arssher) [inline safekeeper/src/pull_timeline.rs:215]: Let's assert (bail! if not) that partial segment exists.

Comment (jcsp) [reply safekeeper/src/pull_timeline.rs:215]: Added a bail, and also comment at the source in `ready_for_eviction` to note that pull timeline assumes the current behavior that we do not evict unless there is a partial segment.

File: test_runner/regress/test_wal_acceptor.py
Comment (jcsp) [reply test_runner/regress/test_wal_acceptor.py]: Pulled inside the wait_until in 92309448e

File: test_runner/regress/test_wal_acceptor.py
Comment (jcsp) [reply test_runner/regress/test_wal_acceptor.py]: Put these into a wait_until in 92309448e

File: safekeeper/src/pull_timeline.rs
Comment (jcsp) [reply safekeeper/src/pull_timeline.rs]: With Arseny's help, refined this to what should be a safe behavior that doesn't hog the lock in abfc227a13da01a3b12a45264415e6f4313e89cc -- the logic is now:
1. try_wal_residence_guard
2. acquire WriteGUardSharedState
3. Take a copy of the control file
4. Check if the control file still indicates evicted state: if not, timeline was woken up in the background, bail out so that the API caller can retry.
5. Drop WriteGuardSharedState before doing I/O
6. Attempt S3 copy of partial segments.  If this fails, it indicates timeline was woken up in the background, bail out so that the API caller can retry
7. Send the control file copy from the start.

So now uneviction is explicitly permitted but the result we send should always be consistent because we're sending the control file copy & a copied partial segment that is the one referenced by that copy.  The failure cases should be very rare in practice, and everything that calls our APIs does retries.  It would be nice to improve the error types in future so that we can get more explicit HTTP responses for the "please retry" cases, at the moment most of this API will give 500s for any anyhow::Error.

**Review by arssher — APPROVED** [review]

File: safekeeper/src/pull_timeline.rs
Comment (jcsp) [reply safekeeper/src/pull_timeline.rs]: oops, removed.

**Pull request merged**
